jardin/instrumentation/notifier.py

```diff
@@ -0,0 +1,88 @@
+import os
+import uuid
+import queue
+import atexit
+import threading
+
+from jardin import config as config
+from jardin.instrumentation.base_subscriber import BaseSubscriber
+
+class Notifer:
+  # We set maxsize to avoid leaking memory if the reporting thread is dead for any reason
+  # in report_event we use `put_nowait` to avoid blocking callers if the queue is full
+  _q = queue.Queue(maxsize=4096)
+  _subscribers = {}
+  _should_stop = False
+  _worker_count = 2
+  _workers = []
+
+  @classmethod
+  def is_enabled(self):
+      return os.environ.get("JARDIN_INSTRUMENTATION_ENABLED", "false") == "true"
+
+  @classmethod
+  def subscribe(self, handler):
+      if not self.is_enabled():
+          return
+      subscriber_id = uuid.uuid4()
+      if not issubclass(type(handler), BaseSubscriber):
+          raise TypeError("Handler must be subclass of BaseSubscriber")
+      self._subscribers[subscriber_id] = handler
+      return subscriber_id
+
+  @classmethod
+  def unsubscribe(self, subscriber_id):
+      if not self.is_enabled():
+          return
+      del self._subscribers[subscriber_id]
+
+  @classmethod
+  def report_event(self, event_name, start_time=0, end_time=0, duration=0, tags={}):
+      if not self.is_enabled() or self._should_stop:
+          return
+      try:
+          self._q.put_nowait((event_name, start_time, end_time, duration, tags))
+      except queue.Full:
+          config.logger.error("Failed to report event {}".format(event_name))
+
+  @classmethod
+  def process_queue(self):
+      if not self.is_enabled():
+          return
+      while True:
+          try:
+              event_name, start_time, end_time, duration, tags = self._q.get()
+              if event_name is None:
```

Comment:
could you factor out either a sentinel value? or if you decide to turn your queue items into objects, add a boolean property to indicate whether it's the sentinel/exit signal?